### PR TITLE
Moves sw-precaching over to using requestWrapper

### PR DESCRIPTION
**Pull request by gauntface**
R: @jeffposnick @addyosmani

This is a step towards #145 although I need to spend a little more time thinking about best approach for cache names

### CODE REVIEW

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js [hunk 1]
```diff
@@ -118,10 +119,15 @@ class BaseCacheManager {
       return;
     }
 
-    let response = await fetch(precacheEntry.getNetworkRequest(), {
+    let response = await this._requestWrapper.fetch({
```
Comment (jeffposnick) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: Have you considered using [`RequestWrapper.fetchAndCache`](https://github.com/GoogleChrome/sw-helpers/blob/master/packages/sw-runtime-caching/src/lib/request-wrapper.js#L177) here instead of `RequestWrapper.fetch`? All the "value add" stuff is in `RequestWrapper.fetchAndCache`.

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js [hunk 2]
```diff
@@ -14,6 +14,7 @@ class BaseCacheManager {
   constructor(cacheName) {
     this._entriesToCache = new Map();
     this._cacheName = cacheName;
+    this._requestWrapper = new RequestWrapper({cacheName: this._cacheName});
```
Comment (jeffposnick) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: Rather than always constructing your own `RequestWrapper`, how about allowing developers to optionally pass in their own?

Developers might want to do that to allow them to assign a `BroadcastCacheUpdate` behavior to their `RequestWrapper`, so that they could announce to controlled pages when there's an update to a precached resource. It could also be useful to override the `CacheableResponse` behavior if they have custom criteria for determining whether a precaching response is valid.

One problem I see is figuring out what would happen if the pass in both a `cacheName` and a `RequestWrapper` that already has a differently named cache.

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js (hunk 1 above)
Comment (gauntface) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: There is one problem with that. The current precaching code will fetch request with the cache busting and then cache it against the origin URL to make future cache.matches possible with the original URL. fetchAndCache will just cache the URL with the cache busting.

**Comment by gauntface** [comment]
@jeffposnick In case you the comment gets lost:

1. I can use fetchAndCache and the current code in precache will cache the reponse against the origin request URL (i.e. the URL without the cache busting search param)
2. I've actually moved unrevisioned assets to the default cache from requestwrapper.

I still need to make the cachenames customisable by a developer, but I'll tackle that in a seperate PR.

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js (hunk 2 above)
Comment (addyosmani) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: I hadn't considered the use-case of someone wanting to construct/pass their own Requestwrapper to control pages when there's updates made. Given @gauntface might need to spend time designing how best to handle passing both cacheName and a RequestWrapper that already has a different cache name, perhaps this can be deferred to another PR?

**Comment by addyosmani** [comment]
> I still need to make the cachenames customisable by a developer, but I'll tackle that in a seperate PR.

This SGTM. I left a comment on the first of your commits which may be showing up as outdated, but it looks like you're going to tackle that work in an iteration PR anyway. Full steam ahead!

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js [hunk 3]
```diff
@@ -13,7 +13,8 @@ class BaseCacheManager {
    */
   constructor(cacheName) {
     this._entriesToCache = new Map();
-    this._cacheName = cacheName;
+    this._requestWrapper = new RequestWrapper({cacheName});
```
Comment (jeffposnick) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]:
```js
this._requestWrapper = new RequestWrapper({
  cacheName,
  fetchOptions: {
    credentials: 'same-origin',
  },
});
```

gives you the same [`fetch` options](https://github.com/GoogleChrome/sw-helpers/blob/master/packages/sw-runtime-caching/src/lib/request-wrapper.js#L50) that the previous code (and like `sw-precache`) uses.

**Comment by jeffposnick** [comment]
Just the one additional comment about `fetchOptions`.

Would you like me to modify `RequestWrapper.fetchAndCache` so that it takes in both the request and a cache key URL to support the use case you need? That seems like a straightforward change, and I've created https://github.com/GoogleChrome/sw-helpers/issues/178 to track that if you'd in favor.

It doesn't need to block this PR.

File: packages/sw-precaching/src/lib/controllers/base-cache-manager.js (hunk 3 above)
Comment (gauntface) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: Possible stupid question, is there a reason for this not to be the default option?

Comment (jeffposnick) [inline packages/sw-precaching/src/lib/controllers/base-cache-manager.js]: Definitely not a stupid question. It's not the default for the underlying `fetch()`, so matching `fetch()` behavior by default unless otherwise required seemed like the least-surprising thing to do.

We ran into a scenario with IOWA in which we specifically needed this set, which is how it ended up in `sw-precache` initially.

I'm open to setting different values by default in `RequestWrapper` directly if you'd like to open an issue to discuss.

**Comment by jeffposnick** [comment]
👍
